### PR TITLE
Add Starlette client (#148)

### PR DESCRIPTION
**Pull request by jonathanunderwood**
This commit adds a new Starlette client along with documentation and
tests. The Starlette client is suitable for use with frameworks that
build on top of Starlette such as FastAPI and Bocadillo.

This implements #148.

**What kind of change does this PR introduce?** (check at least one)

- [ ] Bugfix
- [x] Feature
- [ ] Code style update
- [ ] Refactor
- [ ] Other, please describe:

**Does this PR introduce a breaking change?** (check one)

- [ ] Yes
- [x] No

---

- [x] You consent that the copyright of your pull request source code belongs to Authlib's author.

### CODE REVIEW

**Comment by jonathanunderwood** [comment]
@lepture it seems you've refactored a lot in the last week during which time I gained work clearance to submit the work on a starlette client. 

It's unclear to me why `_send_token_update` raises a NotImplementedError, though I suspect this is in response to me pointing out the flask client didn't implement it, and I see you've not implemented it for Flask. So, it looks like this is needed for the Starlette client, but it's not really clear to me how that's supposed to work.  Can you explain?

File: docs/client/starlette.rst [hunk 1]
```diff
@@ -0,0 +1,377 @@
+.. _starlette_client:
+
+Starlette OAuth Client
+===================
+
+.. meta::
+    :description: The built-in Starlette integrations for OAuth 1.0 and
+        OAuth 2.0 clients.
+
+.. module:: authlib.starlette.client
```
Comment (lepture) [inline docs/client/starlette.rst]: here is `authlib.integrations.starlette_client`

**Comment by lepture** [comment]
@jonathanunderwood thanks a lot for your work. I think the most important feature for  Starlette client is **async**. That's why I'm creating a `httpx_client` now. I'll ping you when `httpx_client` is ready. `httpx_client` should contain Async OAuth Client.

File: authlib/integrations/starlette_client/remote_app.py [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+
+
+class RemoteApp(OAuthClient):
+    """Starlette integrated RemoteApp of :class:`~authlib.client.OAuthClient`.
+
+    This has built-in hooks for ``OAuthClient``.
+
+    """
+
+    def __init__(self, name, fetch_token=None, **kwargs):
```
Comment (lepture) [inline authlib/integrations/starlette_client/remote_app.py]: from the latest code, you don't need to rewrite this `__init__`.

**Comment by jonathanunderwood** [comment]
> @jonathanunderwood thanks a lot for your work. I think the most important feature for Starlette client is **async**. That's why I'm creating a `httpx_client` now. I'll ping you when `httpx_client` is ready. `httpx_client` should contain Async OAuth Client.

I agree that would be nice to have. Could we merge this now and migrate the integration  to the new `httpx_client` client when it is available? - this is already usable with the Starlette framework family, even if it is synchronous.

File: docs/client/starlette.rst (hunk 1 above)
Comment (jonathanunderwood) [reply docs/client/starlette.rst]: Fixed.

**Comment by lepture** [comment]
@jonathanunderwood yes, I can merge it for now. But there won't be a new release soon. I hope I can make a new release in next month. Are you going to use the master code in your project?

**Comment by jonathanunderwood** [comment]
> @jonathanunderwood yes, I can merge it for now. But there won't be a new release soon. I hope I can make a new release in next month. Are you going to use the master code in your project?

That's great to hear - it means I won't have to redo the legal review at work at a later date for this specific code. Yes, I'll be using the master branch for our project.

**Comment by lepture** [comment]
@jonathanunderwood everything looks good for me. But if you could provide a more "starlette-like" database documentation, that would be great. The current documentation is using Django's database design. For example, Flask is using SQLAlchemy for the database example.

**Comment by jonathanunderwood** [comment]
I've made the database design section less like the Django one. But, it's to be noted that Starlette has now ORM support at all, and so is agnostic. I am not convinced that section of the documentation really has much value, but it'll be a starting point to improve on.

In all of the client docs, actually, the database sections don't really make a lot of sense. For example, this part:

```
        def to_token(self):
            return dict(
                oauth_token=self.access_token,
                oauth_token_secret=self.alt_token,
            )
```
which refers to attributes that don't exist for that class. This is true for flask and django client docs. If I'm honest, the documentation for the clients could be more helpful if restructured and significantly re-written, but that's out of scope for this PR, I think.

**Comment by lepture** [comment]
@jonathanunderwood I merged it for now.

Do you have any idea on how to restructure the documentation?

**Comment by jonathanunderwood** [comment]
> Do you have any idea on how to restructure the documentation?

Well, the problem with the set of client docs at the moment is that there's a lot of cut+paste between them, with small tweaks. That indicates that it might be better to have a single document for all the clients that communicates the common concepts across the clients, and then some framework specific notes for each framework.

The other big problem with the docs at present is that they are sketches of how the library could be used, rather than directly run-able code. It would be better to have some super simple but complete code for each framework demonstrating how the clients should be used.

The FastAPI documentation is a great example of directly useable documentation.

**Comment by jonathanunderwood** [comment]
Also: many thanks for merging, and your feedback on all of this. And your work on this library.

**Comment by lepture** [comment]
> The other big problem with the docs at present is that they are sketches of how the library could be used, rather than directly run-able code.

Yes, because it is not a tutorial. For now, there is no real tutorial for Authlib. It is good to have a tutorial for each part of Authlib. But tutorial takes much time to write.

I'm still working on this project in my part time. Although I've gotten some customers via https://authlib.org/plans now, it is not enough to support my life right now. 

I will rewrite the documentation again when v0.13 released. With v0.13, there should be only 2 or 3 RFCs left. v0.13 is a huge release, it is the very step towards v1.0.

**Comment by lepture** [comment]
@jonathanunderwood 

> the problem with the set of client docs at the moment is that there's a lot of cut+paste between them, with small tweaks.

I have updated the documentation for web framework integrations:

1. https://docs.authlib.org/en/latest/client/frameworks.html
2. https://docs.authlib.org/en/latest/client/flask.html
3. https://docs.authlib.org/en/latest/client/django.html

Now, Flask and Django share the same documentation now. Does this make sense to you?

Starlette documentation is not updated yet. I'll change the documentation when it becomes "async".

For ready to use examples, I'm going to create example repositories for every framework.

**Comment by jonathanunderwood** [comment]
That sounds great, I'll read over them when I get a chance

**Comment by lepture** [comment]
@jonathanunderwood here is the re-implementation of the starlette client https://github.com/lepture/authlib/pull/155

**Pull request merged**